jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ResourceHandler.java

```diff
@@ -82,126 +56,67 @@
+        {
+            Context context = ContextHandler.getCurrentContext();
+            if (context != null)
+                _resourceBase = context.getBaseResource();
+        }
 
-// TODO        _context = (context == null ? null : context.getContextHandler());
-//        if (_mimeTypes == null)
+// TODO
 //            _mimeTypes = _context == null ? new MimeTypes() : _context.getMimeTypes();
+        if (_mimeTypes == null)
+            _mimeTypes = new MimeTypes();
 
-        _mimeTypes = new MimeTypes();
-        //_contentFactory = new PathContentFactory();
-        // TODO make caching configurable and disabled by default
-        _contentFactory = new CachingContentFactory(new PathContentFactory());
-        _welcomer = new DefaultWelcomer();
+        setupContentFactory();
 
         super.doStart();
     }
 
-    // for testing only
-    HttpContent.ContentFactory getContentFactory()
-    {
-        return _contentFactory;
-    }
-
-    /**
-     * @return Returns the resourceBase.
-     */
-    public Path getBaseResource()
-    {
-        return _baseResource;
-    }
-
-    /**
-     * @return the cacheControl header to set on all static content.
-     */
-    public String getCacheControl()
-    {
-        return _cacheControl.getValue();
-    }
-
-    /**
-     * @return file extensions that signify that a file is gzip compressed. Eg ".svgz"
-     */
-    public List<String> getGzipEquivalentFileExtensions()
-    {
-        return _gzipEquivalentFileExtensions;
-    }
-
-    public MimeTypes getMimeTypes()
+    private void setupContentFactory()
     {
-        return _mimeTypes;
-    }
-
-    /**
-     * @return Returns the stylesheet as a Resource.
-     */
-    public Path getStylesheet()
-    {
-        if (_stylesheet != null)
-        {
-            return _stylesheet;
-        }
-        else
+        HttpContent.ContentFactory contentFactory = new CachingContentFactory(new ResourceContentFactory(_resourceBase, _mimeTypes, _resourceService.getPrecompressedFormats()));
+        _resourceService.setContentFactory(contentFactory);
+        _resourceService.setWelcomeFactory(pathInContext ->
         {
-            if (_defaultStylesheet == null)
+            if (_welcomes == null)
+                return null;
+
+            for (String welcome : _welcomes)
             {
-                _defaultStylesheet = getDefaultStylesheet();
+                // TODO GW: This logic needs to be extensible so that a welcome file may be a servlet (yeah I know it shouldn't
+                //          be called a welcome file then.   So for example if /foo/index.jsp is the welcome file, we can't
+                //          serve it's contents - rather we have to let the servlet layer to either a redirect or a RequestDispatcher to it.
+                //          Worse yet, if there was a servlet mapped to /foo/index.html, then we need to be able to dispatch to it
+                //          EVEN IF the file does not exist.
```

Comment:
this will need to be implemented sooner rather than later

Reply:
Perhaps welcome file at ResourceHandler doesn't support the full welcome file behaviors?
But DefaultServlet absolutely has to.